{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "-------------------- /nfs/software/helmod/modulefiles/Core ---------------------\n",
      "   AmpliconNoise/1.27-fasrc01              gcc/4.8.2-fasrc01\n",
      "   Anaconda/1.9.2-fasrc01                  gcc/4.9.3-fasrc01\n",
      "   Anaconda/2.1.0-fasrc01                  gcc/5.2.0-fasrc02\n",
      "   Anaconda/2.5.0-fasrc01                  gcc/5.3.0-fasrc01\n",
      "   Anaconda/4.3.0-fasrc01           (D)    gcc/6.2.0-fasrc01\n",
      "   Anaconda3/2.1.0-fasrc01                 gcc/7.1.0-fasrc01\n",
      "   Anaconda3/4.3.0-gcb01                   gcc/7.3.0-gcb01                 (D)\n",
      "   Anaconda3/5.0.1-gcb01                   gcta/1.26.0-gcb01\n",
      "   Anaconda3/2019.10-gcb01                 gd/2.0.28-fasrc01\n",
      "   Anaconda3/2019.10-gcb02          (D)    gdal/1.11.1-fasrc01\n",
      "   BEAST/2.4.2-gcb01                       gdal/2.4.1-gcb01                (D)\n",
      "   BSMAP/2.90-gcb01                        ghc/7.6.1-fasrc01\n",
      "   CLAPACK/3.2.1-gcb01                     ghc/7.8.3-fasrc01               (D)\n",
      "   GATK/4.0.2.1-gcb01                      glib/2.20.4-fasrc01\n",
      "   GATK/4.1.0.0-gcb01                      glib/2.32.4-fasrc01             (D)\n",
      "   GATK/4.1.0.0-gcb02                      glibc/2.14-gcb01\n",
      "   GATK/4.1.3.0-gcb01               (D)    glibc/2.17-gcb01\n",
      "   GenomeAnalysisTK/3.1.1-gcb01            glibc/2.18-gcb01\n",
      "   GenomeAnalysisTK/3.5-gcb01              glibc/2.20-gcb01                (D)\n",
      "   GenomeAnalysisTK/3.7-gcb01              gmp/5.1.3-fasrc01\n",
      "   GenomeAnalysisTK/4.0.2.1-fasrc01        gmp/6.0.0-fasrc01\n",
      "   GenomeAnalysisTK/4.0.2.1-gcb01          gmp/6.0.0-fasrc02\n",
      "   GenomeAnalysisTK/4.1.0.0-gcb01          gmp/6.1.0-fasrc01\n",
      "   GenomeAnalysisTK/4.1.0.0-gcb02          gmp/6.1.1-fasrc02\n",
      "   GenomeAnalysisTK/4.1.3.0-gcb01   (D)    gmp/6.1.2-fasrc01               (D)\n",
      "   HISAT2/2.1.0-gcb01                      gnuplot/4.6.4-fasrc02\n",
      "   HTSeq/0.6.1-fasrc01                     golang/1.10.3-gcb01\n",
      "   IGV/2.3.61-gcb01                        gperf/3.0.4-fasrc01\n",
      "   ImageMagick/7.0.1-gcb01                 gperf/3.0.4-fasrc02             (D)\n",
      "   JAGS/4.1.0-fasrc01                      gsl/1.16-fasrc02\n",
      "   MACS/1.4.2-fasrc01                      gsl/1.16-fasrc03\n",
      "   MUMmer/3.23-gcb01                       gsl/2.5-fasrc01                 (D)\n",
      "   MaSuRCA/3.2.3-gcb01                     gstreamer/0.10.30-fasrc01\n",
      "   OpenBLAS/0.2.20-gcb01                   gtool/0.7.5-gcb01\n",
      "   PBSuite/15.8.24-gcb01                   hdf5/1.8.12-fasrc04\n",
      "   ParsInsert/1.04-fasrc01                 hdf5/1.8.12-fasrc08\n",
      "   R/3.1.0-fasrc01                         hdf5/1.8.17-fasrc02             (D)\n",
      "   R/3.2.0-gcb01                           htslib/1.1-fasrc01\n",
      "   R/3.2.2-fasrc03                         htslib/1.2.1-gcb01\n",
      "   R/3.3.0-fasrc01                         htslib/1.3.1-gcb01\n",
      "   R/3.3.1-gcb01                           htslib/1.9-fasrc01              (D)\n",
      "   R/3.4.0-gcb01                           hyphy/20161104-gcb01\n",
      "   R/3.4.2-fasrc01                         icu4c/54.1-fasrc01\n",
      "   R/3.5.1-gcb01                           infernal/1.0.2-fasrc01\n",
      "   R/3.5.3-gcb01                           java/1.7.0_60-fasrc01\n",
      "   R/3.5.3-gcb02                           java/1.8.0_45-fasrc01\n",
      "   R/3.6.1-gcb01                           java/9-gcb01                    (D)\n",
      "   R/3.6.1-gcb02                           jdk/1.7.0_60-fasrc01\n",
      "   R/3.6.1-gcb03                    (D)    jdk/1.8.0_45-fasrc01\n",
      "   R_core/3.1.0-fasrc01                    jdk/9-gcb01                     (D)\n",
      "   R_core/3.2.2-fasrc02                    jpeg/6b-fasrc02\n",
      "   R_core/3.2.2-fasrc03                    kallisto/0.42.4-gcb01\n",
      "   R_core/3.3.0-fasrc01                    kentUtils/v302-gcb01\n",
      "   R_core/3.3.1-gcb01                      lapack/3.6.1-gcb01\n",
      "   R_core/3.4.0-gcb01                      libffi/3.2.1-fasrc01\n",
      "   R_core/3.4.2-fasrc01                    libpng/1.5.21-fasrc01\n",
      "   R_core/3.5.1-gcb01                      libpng/1.6.23-fasrc01           (D)\n",
      "   R_core/3.5.3-gcb01                      libtool/2.4.6-gcb01\n",
      "   R_core/3.5.3-gcb02                      libxml2/2.7.8-fasrc01\n",
      "   R_core/3.6.1-gcb01                      libxml2/2.7.8-fasrc02           (D)\n",
      "   R_core/3.6.1-gcb02                      llvm/3.5.1-fasrc02\n",
      "   R_core/3.6.1-gcb03               (D)    lynx/2.8.8-gcb01\n",
      "   R_packages/3.1.0-fasrc01                macs2/2.1.0.20140616-fasrc01\n",
      "   R_packages/3.2.2-fasrc03                macs2/2.1.2.20181108-gcb01      (D)\n",
      "   R_packages/3.3.0-fasrc01                matlab/R2014b-gcb01\n",
      "   R_packages/3.3.0-fasrc02                matlab/R2015a-gcb01\n",
      "   R_packages/3.3.1-gcb01                  matlab/R2017b-gcb01\n",
      "   R_packages/3.4.0-gcb01                  matlab/R2018b-gcb01\n",
      "   R_packages/3.4.2-fasrc01                matlab/R2020a-fasrc01           (D)\n",
      "   R_packages/3.5.1-gcb01                  mbuffer/20180625-gcb01\n",
      "   R_packages/3.5.3-gcb01                  meraculous/2.0.5-gcb01\n",
      "   R_packages/3.5.3-gcb02                  mongodb/2.7.4-fasrc01\n",
      "   R_packages/3.6.1-gcb01                  mongodb/3.2.8-gcb01             (D)\n",
      "   R_packages/3.6.1-gcb02                  mothur/1.25.0-fasrc01\n",
      "   R_packages/3.6.1-gcb03           (D)    mpc/1.0.1-fasrc01\n",
      "   Rmpi/0.6.6-gcb01                        mpc/1.0.2-fasrc01\n",
      "   Rmpi/0.6.9-gcb01                 (D)    mpc/1.0.2-fasrc03\n",
      "   STAR/2.5.0c-fasrc01                     mpc/1.0.3-fasrc01\n",
      "   STAR/2.7.2b-gcb01                (D)    mpc/1.0.3-fasrc04\n",
      "   SeqPrep/1.1-fasrc01                     mpc/1.0.3-fasrc06               (D)\n",
      "   StringTie/2.0.5-gcb01                   mpfr/3.1.2-fasrc01\n",
      "   StringTie/2.1.1-gcb01            (D)    mpfr/3.1.2-fasrc02\n",
      "   TrimGalore/0.4.1-gcb01                  mpfr/3.1.3-fasrc01\n",
      "   TrimGalore/0.6.5-fasrc01         (D)    mpfr/3.1.4-fasrc02\n",
      "   VEP/84-gcb01                            mpfr/3.1.5-fasrc01              (D)\n",
      "   autoconf/2.69-fasrc01                   muscle/3.8.31-fasrc01\n",
      "   automake/1.15.1-gcb01                   ncbi-blast/2.2.22-fasrc01\n",
      "   bazel/20160330-fasrc01                  ncbi-blast/2.3.0+-gcb01         (D)\n",
      "   bcftools/1.2-fasrc01                    netcdf/4.3.2-fasrc03\n",
      "   bcftools/1.3.1-gcb01                    node/8.11.3-gcb01\n",
      "   bcftools/1.10.2-fasrc01          (D)    openmpi/1.10.7-gcb01\n",
      "   bcl2fastq2/v2.20.0.422-gcb01            openmpi/3.1.3-gcb01             (D)\n",
      "   bcl2fastq2/2.15.0.4-fasrc02             orc/0.4.24-fasrc01\n",
      "   bcl2fastq2/2.17.1.14-gcb01       (D)    parallel/20160322-fasrc01\n",
      "   beagle/2.1.2-gcb01                      pcre/8.37-fasrc02\n",
      "   bedops/2.4.35-gcb01                     perl/5.10.1-fasrc01\n",
      "   bedtools2/2.19.1-gcb01                  perl/5.10.1-fasrc04             (D)\n",
      "   bedtools2/2.25.0-fasrc01         (D)    perl-modules/5.10.1-fasrc11\n",
      "   binutils/2.26-fasrc02                   phred/020425-gcb01\n",
      "   bioawk/1.0-gcb01                        picard-tools/2.4.1-gcb01\n",
      "   bismark/0.17.0-gcb01                    pigz/2.3.4-gcb01\n",
      "   blasr/20170124-gcb01                    plink/1.07-fasrc01\n",
      "   blasr/20171009-gcb01                    plink/1.90-fasrc01\n",
      "   blasr/20171011-gcb01             (D)    plink/1.90.3.36-gcb01           (D)\n",
      "   blat/3.5-gcb01                          postgresql/9.4.1-fasrc02\n",
      "   boost/1.54.0-fasrc01                    pplacer/1.0.0-fasrc01\n",
      "   boost/1.55.0-fasrc01                    python/2.7.6-fasrc01\n",
      "   boost/1.59.0-gcb01               (D)    python/2.7.8-fasrc02\n",
      "   bowtie/1.1.1-fasrc01                    python/2.7.9-fasrc01\n",
      "   bowtie2/2.2.2-fasrc01                   python/2.7.11-fasrc01\n",
      "   bowtie2/2.2.4-fasrc01                   python/3.4.1-fasrc01\n",
      "   bowtie2/2.3.4.3-gcb01                   python/3.6.3-gcb01\n",
      "   bowtie2/2.3.5.1-gcb01            (D)    python/3.7.4-gcb01              (D)\n",
      "   bwa/0.7.12-gcb01                        qiime/1.9.1-fasrc01\n",
      "   bzip2/1.0.6-fasrc01                     qiime-dependencies/1.9-fasrc01\n",
      "   cairo/1.8.8-fasrc01                     qiime-packages/1.9.1-fasrc01\n",
      "   cairo/1.12.18-fasrc01            (D)    qpy/2.7.10-fasrc01\n",
      "   canu/1.7.1-gcb01                        qt/5.5.1-fasrc01\n",
      "   cd-hit/3.1.1-fasrc01                    qt-creator/3.5.1-fasrc01\n",
      "   cdbfasta/1.0.0-fasrc01                  raxml/7.3.0-fasrc01\n",
      "   cellranger/3.0.1-gcb01                  raxml/8.2.11-fasrc01            (D)\n",
      "   cellranger/3.1.0-gcb01           (D)    readline/6.3-fasrc01\n",
      "   cellranger_atac/1.1.0-gcb01             root/5.34.05-gcb01\n",
      "   cfitsio/3390-fasrc01                    root/6.02.10-fasrc01            (D)\n",
      "   chimeraslayer/2011.05.19-fasrc01        rsem/1.3.2-gcb01\n",
      "   clearcut/1.0.9-fasrc01                  rtax/0.984-fasrc01\n",
      "   cmake/2.8.12.2-fasrc01                  ruby/2.1.1-fasrc01\n",
      "   cmake/3.5.2-fasrc01              (D)    sambamba/0.6.7-gcb01\n",
      "   cmdstan/2.12.0-fasrc01                  samblaster/0.1.24-gcb01\n",
      "   coreutils/8.25-gcb01                    samtools/0.1.19-fasrc01\n",
      "   cuda/7.5-fasrc01                        samtools/1.1-fasrc02\n",
      "   cudnn/7.0-fasrc01                       samtools/1.2-fasrc01\n",
      "   cufflinks/2.2.1-fasrc01                 samtools/1.3.1-gcb01\n",
      "   curl/7.45.0-fasrc01                     samtools/1.9-gcb01\n",
      "   cutadapt/1.8.1-gcb01                    samtools/1.10-gcb01             (D)\n",
      "   cutadapt/1.8.3-gcb01                    scalapack/2.0.2-gcb01\n",
      "   cutadapt/2.3-gcb01               (D)    scons/2.3.6-fasrc01\n",
      "   cwl-utils/0.2-gcb01                     singularity/2.3.1-gcb01\n",
      "   cwltool/1.0.20180711112827-gcb01        singularity/2.4.2-gcb01         (D)\n",
      "   cwltool/1.0.20191022103248-gcb01 (D)    snpEff/4.1-gcb01\n",
      "   cytoscape/2.7.0-fasrc01                 snpEff/4.3i-gcb01               (D)\n",
      "   ddsclient/2.1.2-gcb01                   snptest/2.5.2-gcb01\n",
      "   ddsclient/2.1.3-gcb01                   spark/2.3.0-gcb01\n",
      "   ddsclient/2.1.4-gcb01                   speedseq/0.1.2-gcb03\n",
      "   ddsclient/2.2.0-gcb01                   sqlite/3.12.2-gcb01\n",
      "   ddsclient/2.2.1-gcb01                   sratoolkit/2.7.0-fasrc01\n",
      "   ddsclient/2.3.1-gcb01                   sratoolkit/2.8.1-gcb01\n",
      "   ddsclient/2.3.2-gcb01                   sratoolkit/2.9.6-gcb01\n",
      "   ddsclient/2.3.3-gcb01                   sratoolkit/2.10.8-gcb01         (D)\n",
      "   ddsclient/2.3.4-gcb01                   stampy/1.0.28-fasrc01\n",
      "   ddsclient/2.4.0-gcb01                   szip/2.1-fasrc01\n",
      "   ddsclient/2.4.1-gcb01                   szip/2.1-fasrc02                (D)\n",
      "   ddsclient/2.4.2-gcb01                   tabix/0.2.6-fasrc01\n",
      "   ddsclient/2.4.3-gcb01                   tbb/4.4-fasrc02\n",
      "   ddsclient/2.4.4-gcb01                   tensorflow/20160511-fasrc02\n",
      "   ddsclient/2.4.5-gcb01                   texlive/20170520-gcb01\n",
      "   ddsclient/2.4.6-gcb01                   toil/3.21.0-gcb01\n",
      "   ddsclient/2.4.7-gcb01                   toil/3.24.0-gcb01               (D)\n",
      "   ddsclient/2.4.8-gcb01                   tophat/2.0.13-gcb01\n",
      "   ddsclient/3.0.0-gcb01                   trinityrnaseq/2.0.6-fasrc01\n",
      "   ddsclient/3.1.0-gcb01                   trinityrnaseq/2.6.6-gcb01       (D)\n",
      "   ddsclient/3.1.1-gcb01                   ucsc/20150820-fasrc01\n",
      "   ddsclient/3.2.0-gcb01            (D)    udunits/2.2.26-gcb01\n",
      "   deepTools/2.0-gcb01                     unixODBC/2.3.6-gcb01\n",
      "   deepTools/3.4.3-gcb01            (D)    usearch/5.2.236-fasrc01\n",
      "   drm_tools/1.1.29-gcb01                  usearch/9.0.2132-gcb01          (D)\n",
      "   ea-utils/1.1.2.537-fasrc01              vawk/0.0.1-gcb01\n",
      "   eigen/3.2.2-fasrc01                     vcflib/2015Feb18-fasrc01\n",
      "   emacs/24.3-fasrc01                      vcftools/0.1.11-fasrc01\n",
      "   emboss/6.6.0-fasrc01                    vcftools/0.1.12b-gcb01\n",
      "   expat/2.1.0-fasrc01                     vcftools/0.1.13-gcb01\n",
      "   expat/2.1.0-fasrc03                     vcftools/0.1.15-gcb01           (D)\n",
      "   expat/2.2.7-gcb01                (D)    velvet/1.2.10-fasrc04\n",
      "   fastqc/0.11.2-fasrc01                   virtualenv/12.1.1-fasrc02\n",
      "   fastqc/0.11.5-fasrc01            (D)    virtualenvwrapper/4.6.0-fasrc01\n",
      "   fftw/3.3.5-fasrc01                      xz/5.2.2-fasrc01\n",
      "   fftw/3.3.8-gcb01                 (D)    xz/5.2.2-gcb01                  (D)\n",
      "   flex/2.6.0-fasrc01                      zlib/1.2.8-fasrc02\n",
      "   fontconfig/2.12.0-fasrc01               zlib/1.2.8-fasrc05\n",
      "   freebayes/2015Feb18-fasrc01             zlib/1.2.8-fasrc06\n",
      "   freetype/2.6.3-fasrc01                  zlib/1.2.8-fasrc07              (D)\n",
      "   gcc/4.7.4-fasrc01\n",
      "\n",
      "------------- /nfs/software/helmod/apps/lmod/lmod/modulefiles/Core -------------\n",
      "   lmod/5.9.3    settarg/5.9.3\n",
      "\n",
      "  Where:\n",
      "   (D):  Default Module\n",
      "\n",
      "Use \"module spider\" to find all possible modules.\n",
      "Use \"module keyword key1 key2 ...\" to search for all possible modules matching\n",
      "any of the \"keys\".\n",
      "\n"
     ]
    }
   ],
   "source": [
    "module avail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "module load bedtools2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bedtools v2.25.0\n"
     ]
    }
   ],
   "source": [
    "bedtools --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bedtools: flexible tools for genome arithmetic and DNA sequence analysis.\n",
      "usage:    bedtools <subcommand> [options]\n",
      "\n",
      "The bedtools sub-commands include:\n",
      "\n",
      "[ Genome arithmetic ]\n",
      "    intersect     Find overlapping intervals in various ways.\n",
      "    window        Find overlapping intervals within a window around an interval.\n",
      "    closest       Find the closest, potentially non-overlapping interval.\n",
      "    coverage      Compute the coverage over defined intervals.\n",
      "    map           Apply a function to a column for each overlapping interval.\n",
      "    genomecov     Compute the coverage over an entire genome.\n",
      "    merge         Combine overlapping/nearby intervals into a single interval.\n",
      "    cluster       Cluster (but don't merge) overlapping/nearby intervals.\n",
      "    complement    Extract intervals _not_ represented by an interval file.\n",
      "    subtract      Remove intervals based on overlaps b/w two files.\n",
      "    slop          Adjust the size of intervals.\n",
      "    flank         Create new intervals from the flanks of existing intervals.\n",
      "    sort          Order the intervals in a file.\n",
      "    random        Generate random intervals in a genome.\n",
      "    shuffle       Randomly redistrubute intervals in a genome.\n",
      "    sample        Sample random records from file using reservoir sampling.\n",
      "    spacing       Report the gap lengths between intervals in a file.\n",
      "    annotate      Annotate coverage of features from multiple files.\n",
      "\n",
      "[ Multi-way file comparisons ]\n",
      "    multiinter    Identifies common intervals among multiple interval files.\n",
      "    unionbedg     Combines coverage intervals from multiple BEDGRAPH files.\n",
      "\n",
      "[ Paired-end manipulation ]\n",
      "    pairtobed     Find pairs that overlap intervals in various ways.\n",
      "    pairtopair    Find pairs that overlap other pairs in various ways.\n",
      "\n",
      "[ Format conversion ]\n",
      "    bamtobed      Convert BAM alignments to BED (& other) formats.\n",
      "    bedtobam      Convert intervals to BAM records.\n",
      "    bamtofastq    Convert BAM records to FASTQ records.\n",
      "    bedpetobam    Convert BEDPE intervals to BAM records.\n",
      "    bed12tobed6   Breaks BED12 intervals into discrete BED6 intervals.\n",
      "\n",
      "[ Fasta manipulation ]\n",
      "    getfasta      Use intervals to extract sequences from a FASTA file.\n",
      "    maskfasta     Use intervals to mask sequences from a FASTA file.\n",
      "    nuc           Profile the nucleotide content of intervals in a FASTA file.\n",
      "\n",
      "[ BAM focused tools ]\n",
      "    multicov      Counts coverage from multiple BAMs at specific intervals.\n",
      "    tag           Tag BAM alignments based on overlaps with interval files.\n",
      "\n",
      "[ Statistical relationships ]\n",
      "    jaccard       Calculate the Jaccard statistic b/w two sets of intervals.\n",
      "    reldist       Calculate the distribution of relative distances b/w two files.\n",
      "    fisher        Calculate Fisher statistic b/w two feature files.\n",
      "\n",
      "[ Miscellaneous tools ]\n",
      "    overlap       Computes the amount of overlap from two intervals.\n",
      "    igv           Create an IGV snapshot batch script.\n",
      "    links         Create a HTML page of links to UCSC locations.\n",
      "    makewindows   Make interval \"windows\" across a genome.\n",
      "    groupby       Group by common cols. & summarize oth. cols. (~ SQL \"groupBy\")\n",
      "    expand        Replicate lines based on lists of values in columns.\n",
      "    split         Split a file into multiple files with equal records or base pairs.\n",
      "\n",
      "[ General help ]\n",
      "    --help        Print this help menu.\n",
      "    --version     What version of bedtools are you using?.\n",
      "    --contact     Feature requests, bugs, mailing lists, etc.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "bedtools -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Store the document of bedtools**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "bedtools >& bedtools.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bedtools: flexible tools for genome arithmetic and DNA sequence analysis.\n",
      "usage:    bedtools <subcommand> [options]\n",
      "\n",
      "The bedtools sub-commands include:\n",
      "\n",
      "[ Genome arithmetic ]\n",
      "    intersect     Find overlapping intervals in various ways.\n",
      "    window        Find overlapping intervals within a window around an interval.\n",
      "    closest       Find the closest, potentially non-overlapping interval.\n",
      "    coverage      Compute the coverage over defined intervals.\n"
     ]
    }
   ],
   "source": [
    "head bedtools.log"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
