{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "module load bedops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bedops\n",
      "  citation: http://bioinformatics.oxfordjournals.org/content/28/14/1919.abstract\n",
      "            https://doi.org/10.1093/bioinformatics/bts277\n",
      "  version:  2.4.35 (typical)\n",
      "  authors:  Shane Neph & Scott Kuehn\n",
      "\n",
      "      USAGE: bedops [process-flags] <operation> <File(s)>*\n",
      "\n",
      "          Every input file must be sorted per the sort-bed utility.\n",
      "          Each operation requires a minimum number of files as shown below.\n",
      "            There is no fixed maximum number of files that may be used.\n",
      "          Input files must have at least the first 3 columns of the BED specification.\n",
      "          The program accepts BED and Starch file formats.\n",
      "          May use '-' for a file to indicate reading from standard input (BED format only).\n",
      "\n",
      "      Process Flags:\n",
      "          --chrom <chromosome> Jump to and process data for given <chromosome> only.\n",
      "          --ec                 Error check input files (slower).\n",
      "          --header             Accept headers (VCF, GFF, SAM, BED, WIG) in any input file.\n",
      "          --help               Print this message and exit successfully.\n",
      "          --help-<operation>   Detailed help on <operation>.\n",
      "                                 An example is --help-c or --help-complement\n",
      "          --range L:R          Add 'L' bp to all start coordinates and 'R' bp to end\n",
      "                                 coordinates. Either value may be + or - to grow or\n",
      "                                 shrink regions.  With the -e/-n operations, the first\n",
      "                                 (reference) file is not padded, unlike all other files.\n",
      "          --range S            Pad or shrink input file(s) coordinates symmetrically by S.\n",
      "                                 This is shorthand for: --range -S:S.\n",
      "          --version            Print program information.\n",
      "\n",
      "      Operations: (choose one of)\n",
      "          -c, --complement [-L] File1 [File]*\n",
      "          -d, --difference ReferenceFile File2 [File]*\n",
      "          -e, --element-of [bp | percentage] ReferenceFile File2 [File]*\n",
      "                 by default, -e 100% is used.  'bedops -e 1' is also popular.\n",
      "          -i, --intersect File1 File2 [File]*\n",
      "          -m, --merge File1 [File]*\n",
      "          -n, --not-element-of [bp | percentage] ReferenceFile File2 [File]*\n",
      "                 by default, -n 100% is used.  'bedops -n 1' is also popular.\n",
      "          -p, --partition File1 [File]*\n",
      "          -s, --symmdiff File1 File2 [File]*\n",
      "          -u, --everything File1 [File]*\n",
      "          -w, --chop [bp] [--stagger <nt>] [-x] File1 [File]*\n",
      "                 by default, -w 1 is used with no staggering.\n",
      "      \n",
      "Example: bedops --range 10 -u file1.bed\n",
      "      NOTE: Only operations -e|n|u preserve all columns (no flattening)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "bedops --help > bedops.log\n",
    "cat bedops.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
